### ISSUE TITLE
MinGW/msys2 build with mingw-w64-x86_64-toolchain

### ISSUE THREAD

**Issue by gaaned92** [Exiv2/exiv2#1353]
System W10/MSYS2 GCC 10.2.0

Exiv2 v0.27.3-138-g4ca1f2b9

```
- CMake Generator:   MSYS Makefiles
-- CMAKE_BUILD_TYPE:  Release
-- Compiler info: GNU (C:/msys64/mingw64/bin/g++.exe) ; version: 10.2.0
-- CMAKE_CXX_STANDARD:11
--  --- Compiler flags ---
-- General:           -Wno-deprecated-declarations
         -Wp,-D_GLIBCXX_ASSERTIONS
         -Wall
         -Wcast-align
         -Wpointer-arith
         -Wformat-security
         -Wmissing-format-attribute
         -Woverloaded-virtual
         -W

```

Compilation errors in http.cpp:

```
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp: In function 'int forgive(int, int&)':
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:129:12: error: 'WSAGetLastError' was not declared in this scope
  129 |     err  = WSAGetLastError() ;
      |            ^~~~~~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:134:28: error: 'WSAEWOULDBLOCK' was not declared in this scope; did you mean 'EWOULDBLOCK'?
  134 |     bool bForgive = err == WSAEWOULDBLOCK || err == WSAENOTCONN ;
      |                            ^~~~~~~~~~~~~~
      |                            EWOULDBLOCK
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:134:53: error: 'WSAENOTCONN' was not declared in this scope; did you mean 'ENOTCONN'?
  134 |     bool bForgive = err == WSAEWOULDBLOCK || err == WSAENOTCONN ;
      |                                                     ^~~~~~~~~~~
      |                                                     ENOTCONN
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:135:26: error: 'SOCKET_ERROR' was not declared in this scope
  135 |     bool bError   = n == SOCKET_ERROR ;
      |                          ^~~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp: In function 'int makeNonBlocking(int)':
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:186:3: error: 'ULONG' was not declared in this scope
  186 |   ULONG  ioctl_opt = 1;
      |   ^~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:187:30: error: 'FIONBIO' was not declared in this scope
  187 |   return ioctlsocket(sockfd, FIONBIO, &ioctl_opt);
      |                              ^~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:187:40: error: 'ioctl_opt' was not declared in this scope
  187 |   return ioctlsocket(sockfd, FIONBIO, &ioctl_opt);
      |                                        ^~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:187:10: error: 'ioctlsocket' was not declared in this scope
  187 |   return ioctlsocket(sockfd, FIONBIO, &ioctl_opt);
      |          ^~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp: In function 'int Exiv2::http(Exiv2::Dictionary&, Exiv2::Dictionary&, std::string&)':
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:208:5: error: 'WSADATA' was not declared in this scope
  208 |     WSADATA wsaData;
      |     ^~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:209:16: error: 'MAKEWORD' was not declared in this scope
  209 |     WSAStartup(MAKEWORD(2,2), &wsaData);
      |                ^~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:209:32: error: 'wsaData' was not declared in this scope; did you mean 'wstat'?
  209 |     WSAStartup(MAKEWORD(2,2), &wsaData);
      |                                ^~~~~~~
      |                                wstat
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:209:5: error: 'WSAStartup' was not declared in this scope
  209 |     WSAStartup(MAKEWORD(2,2), &wsaData);
      |     ^~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:256:35: error: 'AF_INET' was not declared in this scope
  256 |     int     sockfd = (int) socket(AF_INET , SOCK_STREAM,IPPROTO_TCP) ;
      |                                   ^~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:256:45: error: 'SOCK_STREAM' was not declared in this scope
  256 |     int     sockfd = (int) socket(AF_INET , SOCK_STREAM,IPPROTO_TCP) ;
      |                                             ^~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:256:57: error: 'IPPROTO_TCP' was not declared in this scope
  256 |     int     sockfd = (int) socket(AF_INET , SOCK_STREAM,IPPROTO_TCP) ;
      |                                                         ^~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:256:28: error: 'socket' was not declared in this scope; did you mean 'sockfd'?
  256 |     int     sockfd = (int) socket(AF_INET , SOCK_STREAM,IPPROTO_TCP) ;
      |                            ^~~~~~
      |                            sockfd
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:263:25: error: aggregate 'Exiv2::http(Exiv2::Dictionary&, Exiv2::Dictionary&, std::string&)::sockaddr_in serv_addr' has incomplete type and cannot be defined
  263 |     struct  sockaddr_in serv_addr   ;
      |                         ^~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:267:35: error: 'inet_addr' was not declared in this scope
  267 |     serv_addr.sin_addr.s_addr   = inet_addr(servername_p);
      |                                   ^~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:269:35: error: 'htons' was not declared in this scope
  269 |     serv_addr.sin_port          = htons(atoi(port_p));
      |                                   ^~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:273:53: error: 'INADDR_NONE' was not declared in this scope
  273 |     if (serv_addr.sin_addr.s_addr == (unsigned long)INADDR_NONE)
      |                                                     ^~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:275:32: error: 'gethostbyname' was not declared in this scope
  275 |         struct hostent* host = gethostbyname(servername_p);
      |                                ^~~~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:277:40: error: invalid use of incomplete type 'struct Exiv2::http(Exiv2::Dictionary&, Exiv2::Dictionary&, std::string&)::hostent'
  277 |         memcpy(&serv_addr.sin_addr,host->h_addr,sizeof(serv_addr.sin_addr));
      |                                        ^~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:275:16: note: forward declaration of 'struct Exiv2::http(Exiv2::Dictionary&, Exiv2::Dictionary&, std::string&)::hostent'
  275 |         struct hostent* host = gethostbyname(servername_p);
      |                ^~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:284:14: error: 'connect' was not declared in this scope
  284 |     server = connect(sockfd, (const struct sockaddr *) &serv_addr, serv_len) ;
      |              ^~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:285:20: error: 'SOCKET_ERROR' was not declared in this scope
  285 |     if ( server == SOCKET_ERROR && WSAGetLastError() != WSAEWOULDBLOCK )
      |                    ^~~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:285:36: error: 'WSAGetLastError' was not declared in this scope
  285 |     if ( server == SOCKET_ERROR && WSAGetLastError() != WSAEWOULDBLOCK )
      |                                    ^~~~~~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:285:57: error: 'WSAEWOULDBLOCK' was not declared in this scope; did you mean 'EWOULDBLOCK'?
  285 |     if ( server == SOCKET_ERROR && WSAGetLastError() != WSAEWOULDBLOCK )
      |                                                         ^~~~~~~~~~~~~~
      |                                                         EWOULDBLOCK
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:300:28: error: 'send' was not declared in this scope
  300 |     while ( sleep_ >= 0 && send(sockfd,buffer,n,0) == SOCKET_ERROR /* && WSAGetLastError() == WSAENOTCONN */ ) {
      |                            ^~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:300:55: error: 'SOCKET_ERROR' was not declared in this scope
  300 |     while ( sleep_ >= 0 && send(sockfd,buffer,n,0) == SOCKET_ERROR /* && WSAGetLastError() == WSAENOTCONN */ ) {
      |                                                       ^~~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:301:9: error: 'Sleep' was not declared in this scope; did you mean 'sleep'?
  301 |         Sleep(snooze) ;
      |         ^~~~~
      |         sleep
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:306:114: error: 'WSAGetLastError' was not declared in this scope
  306 |         return error(errors,"error - timeout connecting to server = %s port = %s wsa_error = %d",servername,port,WSAGetLastError());
      |                                                                                                                  ^~~~~~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:315:15: error: 'recv' was not declared in this scope
  315 |     n=forgive(recv(sockfd,buffer,(int)buff_l,0),err) ;
      |               ^~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:382:13: error: 'Sleep' was not declared in this scope; did you mean 'sleep'?
  382 |             Sleep(snooze) ;
      |             ^~~~~
      |             sleep
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:390:21: error: 'WSAGetLastError' was not declared in this scope
  390 |                 ,   WSAGetLastError()
      |                     ^~~~~~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:401:113: error: 'WSAGetLastError' was not declared in this scope
  401 |             return error(errors,"error - no response from server = %s port = %s wsa_error = %d",servername,port,WSAGetLastError());
      |                                                                                                                 ^~~~~~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:407:5: error: 'closesocket' was not declared in this scope
  407 |     closesocket(server) ;
      |     ^~~~~~~~~~~
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp: At global scope:
D:/PROGRAMMES/exiv2source/exiv2/src/http.cpp:127:12: warning: 'int forgive(int, int&)' defined but not used [-Wunused-function]
  127 | static int forgive(int n,int& err)
```

**Comment by clanmills** [comment]
Thanks for report this.

I haven't built Exiv2 with GCC 10 on any platform yet.  Looks is though something is hiding <winsock2.h> from MinGW/msys2.  I'll investigate.

**Comment by piponazo** [comment]
This might be related with the changes I did in #1349 . I can also try to reproduce the windows on my machine with W10/MSYS2 GCC 10.2.0 and take care of solving the issue.

**Comment by clanmills** [comment]
Thanks, @piponazo.  I'll leave this to you.

**Comment by piponazo** [comment]
Could you please share the MSYS packages used and how you call the `cmake` command the first time? In that way, I can try to reproduce the problem on my system.

**Comment by gaaned92** [comment]
@piponazo 
_Could you please share the MSYS packages used_
Do you need the versions of those packages?.All packages are in a very recent versions a few days ago.

_how you call the cmake command the first time_

cmake -G "MSYS Makefiles" -DEXIV2_ENABLE_WIN_UNICODE=On -DCMAKE_CXX_STANDARD=11 -DCMAKE_CXX_FLAGS=-Wno-deprecated  -DCMAKE_INSTALL_PREFIX=/d/programmes/exiv2source/install$BD  -DCMAKE_INSTALL_DATAROOTDIR=/d/programmes/exiv2source/install$BD/share ../exiv2 -DCMAKE_BUILD_TYPE=Release

**Comment by clanmills** [comment]
Ah.  I think it might be wrong to use -G "MSYS Makefiles".   I always use the default -G (which I think is Unix Makefiles).

**Comment by clanmills** [comment]
I'm doing the builds now.  My cmake (which I believe was installed by pacman as documented in README.md) doesn't support "MSYS Makefiles".   There's a story lurking in my mind about "MSYS Makefiles".  I'll have to build cmake from source.  The last time I tried to build cmake from source on MinGW/msys2, it wouldn't compile.  #1224 

I have GCC 7.4.0 on the my build machine, I'll grade to GCC 10 and see what happens.

Can you look at the steps I documented in 5.3 of README.md to set up the dependences and build.

**Comment by piponazo** [comment]
I asked exactly to determine which compiler it was being used. I noticed that in MSYS you can install either `gcc` or `mingw-w64-<i686|x86_64>-toolchain`. The same happens with `cmake`. The default `cmake` package does not have a generator  "MSYS Makefiles" while the MinGW one has it.  Note that when the MinGW packages are installed, a new "MSYS 2 MinGW 64-bits" shortcut is created in the windows menus.
* default gcc version is: 9.3.0
* MinGW gcc version is : 10.2.0

When using the "normal" `gcc` , I have found some compilation errors of "UNIX" stuff not being found. For example:

```bash
[ 24%] Building CXX object src/CMakeFiles/exiv2lib.dir/FileIo.cpp.o
/home/luis/exiv2/src/FileIo.cpp: In member function ‘virtual Exiv2::byte* Exiv2::FileIo::mmap(bool)’:
/home/luis/exiv2/src/FileIo.cpp:329:73: error: ‘fileno’ was not declared in this scope; did you mean ‘FileIo’?
  329 |         void* rc = ::mmap(nullptr, p_->mappedLength_, prot, MAP_SHARED, fileno(p_->fp_), 0);
```

When using the MinGW gcc, I get different errors at different points:

```bash
[  1%] Building CXX object xmpsdk/CMakeFiles/exiv2-xmp.dir/src/XMPUtils.cpp.obj
C:/dev/msys64/home/luis/exiv2/xmpsdk/src/XMPUtils.cpp: In static member function 'static void XMPUtils::CurrentDateTime(XMP_DateTime*)':
C:/dev/msys64/home/luis/exiv2/xmpsdk/src/XMPUtils.cpp:74:25: error: 'localtime_r' was not declared in this scope; did you mean 'localtime_s'?
   74 |  #define ansi_localtime localtime_r
```

Actually the 2nd case is more worrying for me. I would spend time in fixing such compilation errors in the Exiv2 codebase, but the failure we saw in the latter snippet are about XMPSDK. We are more restrictive about modifying things in that folder because it is 3rd party code. We would need to add something like this to support the MinGW case in XMPSDK:

```
#ifdef __GNUC__
#ifdef __MINGW32__
```

@gaaned92 Did you see any of those errors? The more details you give about your use-case the better for us to understand your needs. I would also like to know if you are just trying to compile Exiv2 with MSYS "for fun" or because it is needed as a dependency in other project.

**Comment by gaaned92** [comment]
When installing MSYS2, you will get 2 or 3 shells: MSYS shell, Mingw32 shell (if you want build 32 bits apps) and MINGW64 shell to build 64 bits apps.
MSYS shell should be only used for MSYS2 maintenance and update. (pacman...)

For building application you must exclusively use MINGW64 shell that provides the correct environment.

That is to say that I only use MINGW64 shell to build apps.

So there is no default or normal gcc. It comes from erroneous use of MSYS2.
To get the NORMAL mingw64 shell you can create a shortcut to C:\msys64\mingw64.exe.

**Comment by clanmills** [comment]
Thank You for your work and contributions to RT and open-source.  I have somebody working with me to get legal advice concerning adding CR3 (and other ISOBMFF container) support.   I hope to a release of Exiv2 in 2021 will have CR3 support:  https://github.com/Exiv2/exiv2/issues/1229#issuecomment-705350266

-------------

We need to get on the same page here!  Are you cross compiling for MinGW from Linux?

When I first started working with MinGW/msys2, the installer created a little GUI that enabled you to change between the three modes you've described.  Since v0.27 (2 years ago), I've been following the instructions that I have in README.md to set up MinGW/msys2.

This afternoon, I downloaded the latest from https://repo.msys2.org/distrib/x86_64/msys2-x86_64-20200903.exe  That installed GCC 9.3 which suffers from this https://github.com/Exiv2/exiv2/pull/1175#issuecomment-619040336  You'll notice that I said:

> I expect we'll be visiting GCC 9.3.0 on Cygwin and MSYS again soon.

I was unable to persuade pacman to install GCC 10.2

In README.md, I've recommended installing the tools and dependencies with:

```bash
$ for i in base-devel git cmake coreutils python3 man gcc gdb \
    make dos2unix tar diffutils zlib-devel libexpat-devel \
    libiconv-devel gettext-devel
do
  (echo y|pacman -S $i)
done
```

From @piponazo's comments, I think the if I replace `gcc gdb` with `mingw-w64-x86_64-toolchain`, that will install GCC 10.2.  However the pacman installed cmake doesn't support "MinGW Makefiles".

We have to get on the same page here and use the same setup to build MinGW/msys2

**Comment by gaaned92** [comment]
_We have to get on the same page here and use the same setup to build MinGW/msys2_

Yes, after installing MSYS2, you should install additional packages as follows:
- additional msys packages
`pacman -S base-devel  tar git diffutils man`

- mingw64 packages tools and dependencies
`pacman -S mingw-w64-x86_64-{toolchain,cmake,gettext,libexpat,libiconv}`
perhaps some mingw64 packages are missing but I cannot test.

 Packages required to build  an app are **mingw** packages and NOT msys packages.

**Comment by clanmills** [comment]
Thank You @gaaned92 .  Teamwork really works.  Don't know why it's never used in the office.

The fix is:

```
58 MSYS rmills@rmillsmm-w10:~/gnu/github/exiv2/0.27-maintenance/build $ git diff
diff --git a/src/http.cpp b/src/http.cpp
index fc274b7..f8c16d6 100644
--- a/src/http.cpp
+++ b/src/http.cpp
@@ -18,7 +18,11 @@
  * Foundation, Inc., 51 Franklin Street, 5th Floor, Boston, MA 02110-1301 USA.
  */

-#if defined(_MSC_VER) || defined(__CYGWIN__) || defined(__MINGW__)
+#if (defined(__MINGW64__) || defined(__MINGW32__)) && ! defined(__MINGW__)
+#define __MINGW__
+#endif
+
+#if defined(_MSC_VER) || defined(__CYGWIN__) || defined(__MINGW__)
 #define __USE_W32_SOCKETS
 #include <winsock2.h>
 #endif
59 MSYS rmills@rmillsmm-w10:~/gnu/github/exiv2/0.27-maintenance/build $  
```

We demoted the include of <winsock2.h> from config.h.  It is essential in http.cpp.  \_\_MINGW\_\_ is set when using the MSYS tools.  The mingw64 tools have both \_\_MINGW32\_\_ and \_\_MINGW64\_\_, however they don't set \_\_MINGW\_\_.

Later today, I will submit a PR and update our build instructions in README.md to specify the mingw64 tools.  The command to configure the tools and dependencies is:
```bash
for i in base-devel git coreutils dos2unix tar diffutils                          \
    mingw-w64-x86_64-toolchain mingw-w64-x86_64-gcc      mingw-w64-x86_64-gdb     \
    mingw-w64-x86_64-cmake     mingw-w64-x86_64-gettext  mingw-w64-x86_64-python3 \
    mingw-w64-x86_64-libexpat  mingw-w64-x86_64-libiconv mingw-w64-x86_64-zlib;   \
    mingw-w64-x86_64-gtest
do (echo y | pacman -S $i) ; done
```
Additionally, the script msys64.bat requires /mingw64/bin on the PATH.

```cmd
set "PATH=c:\msys64\mingw64;c:\msys64\usr\bin;c:\msys64\usr\local\bin;"
```

If you're happy with this fix, I invite you to close this ticket.

**Comment by piponazo** [comment]
Thanks @gaaned92 for the feedback about how to use MSYS correctly. I managed to compile the exiv2lib latest version at 0.27-maintenance without issues. Maybe the version you are trying to compile (Exiv2 v0.27.3-138-g4ca1f2b9) does not contain the latest change we merged recently in PR #1349.

However I got some linking errors at the point of trying to generate the `exiv2.exe` executable. This is a known issue about `BasicError:setMsgEv`. @clanmills did you have problems about this in your experiments?
```
[ 49%] Linking CXX executable ../bin/exiv2.exe
C:/dev/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles/exiv2.dir/objects.a(params.cpp.obj):params.cpp:(.text$_ZN5Exiv210BasicErrorIcEC1INSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEENS_9ErrorCodeERKT_[_ZN5Exiv210BasicErrorIcEC1INSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEENS_9ErrorCodeERKT_]+0xbd): undefined reference to `__imp__ZN5Exiv210BasicErrorIcE6setMsgEv'
collect2.exe: error: ld returned 1 exit status
```

I also had to comment out some sections of `cmake/compilerFlags.cmake` to manage to compile the code with MSYS2/MinGW64. It looks like some of the `-fstack-***` and `-Wp -D_FORTIFY_SOURCE=2` compiler flags do not play well with this environment. I actually think that all those "special flags" are giving us more headaches than benefits. I'll open another issue to discuss about it.

**Comment by clanmills** [comment]
@piponazo  Everything seemed fine to me.  I used a brand new installation of msys2 and clone of exiv2 0.27-maintenance.

I'm currently creating the PR and will reinstall msys2 on the build server to be sure my changes to README.md are solid.

I'm puzzled by `BasicError:setMsgEv`.  I haven't seen that for more than year.

**Comment by piponazo** [comment]
I also had a brand new installation of MSYS2 and a fresh clone of exiv2 0.27-maintenance ... I'll give it another try when you publish the PR, just to double check your documentation 😉

**Comment by clanmills** [comment]
That would be wonderful.  Thank You @piponazo

**Comment by clanmills** [comment]
I suspect that using the `mingw-w64-x86_64-toolchain` will solve: #1224 because that was being blocked by conan wanting to use  -G "MSYS Makefiles".  I couldn't find/build a version of cmake with that generator.  And I think that will solve building cmake on MinGW.

The fog clears.  The sun shines.  Life is good.

A new task is to get the CI to build this and cygwin64.

**Comment by piponazo** [comment]
According to this link, Appveyor counts with MSYS and Cygwin, so we should be able to make new builds for such environments:

https://www.appveyor.com/docs/windows-images-software/#mingw-msys-cygwin

That's something I can also take care of 😉 . If you prepare a PR with the instructions to build with MSYS2/MinGW64, I will collaborate in the same PR to add the CI job into Appveyor. In that way, we can validate the changes in the same own PR.

**Comment by clanmills** [comment]
That's a deal.  I have a music lesson this morning.  If I can't finish the PR before the teacher arrives, I'll do it after lunch.

**Comment by gaaned92** [comment]
Happy to help you.
I will try to test the patch or PR if done  this afternoon.
I made some trial with appveyor and was able to build rawtherapee with MSYS2. What I was missing was to trigger the build when a commit was done on whatever branch. It's old now and I cannot really help.

And @clanmills, enjoy your music lesson, if I remember euphonium.

**Comment by clanmills** [comment]
> if I remember euphonium.

Correct.  Well done.

I'll do the PR after lunch.
![RobinEuphonium](https://user-images.githubusercontent.com/529982/95845869-9e505700-0d42-11eb-8f35-3b8ea98b15b5.jpg)

**Comment by gaaned92** [comment]
Thank you guys for this fast correction. Closing now
